saleor/graphql/order/mutations/orders.py

```diff
@@ -574,3 +607,219 @@ def perform_mutation(cls, root, info, **data):
+
+
+class OrderLineDelete(EditableOrderValidationMixin, BaseMutation):
+    order = graphene.Field(Order, description="A related order.")
+    order_line = graphene.Field(
+        OrderLine, description="An order line that was deleted."
+    )
+
+    class Arguments:
+        id = graphene.ID(description="ID of the order line to delete.", required=True)
+
+    class Meta:
+        description = "Deletes an order line from an order."
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = OrderError
+        error_type_field = "order_errors"
+
+    @classmethod
+    def perform_mutation(cls, _root, info, id):
+        line = cls.get_node_or_error(
+            info,
+            id,
+            only_type=OrderLine,
+        )
+        order = line.order
+        cls.validate_order(line.order)
+
+        db_id = line.id
+        line_info = OrderLineData(
+            line=line,
+            quantity=line.quantity,
+            variant=line.variant,
+            warehouse_pk=line.allocations.first().stock.warehouse.pk,
+        )
+        delete_order_line(line_info)
+        line.id = db_id
+
+        # Create the removal event
+        events.order_removed_products_event(
+            order=order, user=info.context.user, order_lines=[(line.quantity, line)]
+        )
+
+        recalculate_order(order)
+        return OrderLineDelete(order=order, order_line=line)
+
+
+class OrderLineUpdate(EditableOrderValidationMixin, ModelMutation):
+    order = graphene.Field(Order, description="Related order.")
+
+    class Arguments:
+        id = graphene.ID(description="ID of the order line to update.", required=True)
+        input = OrderLineInput(
+            required=True, description="Fields required to update an order line."
+        )
+
+    class Meta:
+        description = "Updates an order line of an order."
+        model = models.OrderLine
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = OrderError
+        error_type_field = "order_errors"
+
+    @classmethod
+    def clean_input(cls, info, instance, data):
+        instance.old_quantity = instance.quantity
+        cleaned_input = super().clean_input(info, instance, data)
+        cls.validate_order(instance.order)
+
+        quantity = data["quantity"]
+        if quantity <= 0:
+            raise ValidationError(
+                {
+                    "quantity": ValidationError(
+                        "Ensure this value is greater than 0.",
+                        code=OrderErrorCode.ZERO_QUANTITY,
+                    )
+                }
+            )
+        return cleaned_input
+
+    @classmethod
+    def save(cls, info, instance, cleaned_input):
+        line_info = OrderLineData(
+            line=instance,
+            quantity=instance.quantity,
+            variant=instance.variant,
+            warehouse_pk=instance.allocations.first().stock.warehouse.pk,
```

Comment:
If you perform this mutation for draft order then the line doesn't have any allocation. We should fetch `warehouse_pk` only for unconfirmed orders.